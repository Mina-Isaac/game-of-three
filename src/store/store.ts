import { configureStore, combineReducers } from "@reduxjs/toolkit";
import playerReducer from "./playerDetailsSlice";
import gameReducer from "./gameSlice";
import {
  createStateSyncMiddleware,
  initStateWithPrevTab,
  withReduxStateSync,
} from "redux-state-sync";

const stateSyncMiddleware = createStateSyncMiddleware();

export const rootReducer = combineReducers({
  player: playerReducer,
  game: gameReducer,
});

const reducer = withReduxStateSync(rootReducer);

export const store = configureStore({
  reducer,
  middleware: [stateSyncMiddleware],
  devTools: process.env.NODE_ENV === 'development'
});
initStateWithPrevTab(store);

export type RootState = ReturnType<typeof rootReducer>;
