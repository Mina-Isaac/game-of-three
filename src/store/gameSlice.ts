import { createSlice, PayloadAction } from "@reduxjs/toolkit";
import { PlayerState } from "./playerDetailsSlice";
import { RootState } from "./store";

interface GameState {
  startingNumber: number | undefined;
  baseNumber: number | undefined;
  validMove: -1 | 0 | 1 | undefined;
  vs: "human" | "cpu" | undefined;
  lastPlayer: keyof PlayerState;
  canPlay: { player1: boolean; player2: boolean };
  winner: keyof PlayerState | undefined;
}

export const initialState: GameState = {
  startingNumber: undefined,
  baseNumber: undefined,
  validMove: undefined,
  vs: undefined,
  lastPlayer: "player1",
  canPlay: { player1: false, player2: true },
  winner: undefined,
};

export const gameSlice = createSlice({
  name: "game",
  initialState,
  reducers: {
    initializeGame: (state, action: PayloadAction<number>) => {
      const { payload } = action;
      state.startingNumber = payload;
      state.baseNumber = payload;
      state.lastPlayer = "player1";
      state.canPlay = { player1: false, player2: true };
      if (payload % 3 === 0) state.validMove = 0;
      else if (payload % 3 === 1) state.validMove = -1;
      else state.validMove = 1;
    },
    playNextMove: (state) => {
      const newBase = (state.baseNumber! + state.validMove!) / 3;
      state.baseNumber = newBase;
      if (newBase % 3 === 0) state.validMove = 0;
      else if (newBase % 3 === 1) state.validMove = -1;
      else state.validMove = 1;
      const { player1, player2 } = state.canPlay;
      state.canPlay = { player1: !player1, player2: !player2 };
      state.lastPlayer =
        (state.lastPlayer === "player1" && "player2") || "player1";
      if (newBase === 1) state.winner = state.lastPlayer;
    },
    setVs: (state, action: PayloadAction<"human" | "cpu" | undefined>) => {
      state.vs = action.payload;
    },
    startNewGame: (state) => {
      state.baseNumber = undefined;
      state.validMove = undefined;
      state.winner = undefined;
    },
  },
});

export const {
  initializeGame,
  playNextMove,
  setVs,
  startNewGame,
} = gameSlice.actions;

// Selectors
export const selectStartingNumber = (state: RootState) =>
  state.game.startingNumber;
export const selectBaseNumber = (state: RootState) => state.game.baseNumber;
export const selectVs = (state: RootState) => state.game.vs;
export const selectValidMove = (state: RootState) => state.game.validMove;
export const selectLastPlayer = (state: RootState) => state.game.lastPlayer;
export const SelectWinner = (state: RootState) => state.game.winner;
export const selectCanPlay = (palyeKey: keyof PlayerState | undefined) => (
  state: RootState
) => {
  if (!palyeKey) return false;
  return state.game.canPlay[palyeKey];
};

export default gameSlice.reducer;
