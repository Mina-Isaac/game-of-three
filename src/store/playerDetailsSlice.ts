import { createSlice, PayloadAction } from "@reduxjs/toolkit";
import { RootState } from "./store";

export interface Player {
  type: "human" | "cpu";
  name: string;
}
export interface PlayerState {
  player1: undefined | Player;
  player2: undefined | Player;
}

export const initialState: PlayerState = {
  player1: undefined,
  player2: undefined,
};

export const playerSlice = createSlice({
  name: "players",
  initialState,
  reducers: {
    initializePlayer: (
      state,
      action: PayloadAction<Player & { playerKey: keyof PlayerState }>
    ) => {
      const { playerKey, ...details } = action.payload;
      state[playerKey] = { ...details };
    },
  },
});

export const { initializePlayer } = playerSlice.actions;

// Selectors
export const selectPlayer1 = (state: RootState) => state.player.player1;
export const selectPlayer2 = (state: RootState) => state.player.player2;

export default playerSlice.reducer;
