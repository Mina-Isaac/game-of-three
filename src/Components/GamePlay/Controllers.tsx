import React, { useState } from "react";
import { useDispatch, useSelector } from "react-redux";
import styled from "styled-components";
import {
  playNextMove,
  selectCanPlay,
  selectValidMove,
} from "../../store/gameSlice";
import { PlayerState } from "../../store/playerDetailsSlice";

const Wrapper = styled.div`
  position: relative;
  display: flex;
  justify-content: space-between;
  align-items: flex-end;
  margin-top: auto;
  width: 100%;
  padding: 10px 24px 10px;
  border-top: 1px solid #00000069;
`;

const Div = styled.div`
  display: inline-block;
  height: fit-content;
  width: fit-content;
  border-radius: 50%;

  :hover button:enabled {
    background-color: #0e78e2cc;
    height: 75px;
    width: 75px;
    margin-bottom: -13px;
    ${(props: { edge?: "right" | "left" }) => `margin-${props.edge}:-13px`};
  }
`;

const Button = styled.button`
  height: 62px;
  width: 62px;
  border: 1px solid rgb(0 0 0);
  font-size: xx-large;
  border-radius: 50%;
  background-color: #50aadd9c;
  transition: all 130ms;
  :disabled {
    background-color: #e624244a;
  }
`;

const Popup = styled.div`
  width: 154px;
  height: 39px;
  border-radius: 19px;
  border: 1px solid black;
  position: absolute;
  top: -45%;
  left: 26%;
  background-color: rgb(226 13 13 / 70%);
  z-index: 1;
  display: ${(props: { show: boolean }) => (props.show ? "flex" : "none")};
  justify-content: center;
  align-items: center;
  padding: 2%;
  color: white;
  font-size: x-large;
  font-weight: bold;
`;
interface Props {
  playingAs: keyof PlayerState | undefined;
}

const Controllers: React.FC<Props> = ({ playingAs }) => {
  const dispatch = useDispatch();
  const validMove = useSelector(selectValidMove);
  const canPlay = useSelector(selectCanPlay(playingAs));
  const [showPopup, setShowPopup] = useState(false);

  const handleClick = (value: -1 | 0 | 1) => {
    if (!canPlay) return;
    if (value === validMove) dispatch(playNextMove());
    else {
      setShowPopup(true);
      setTimeout(() => setShowPopup(false), 400);
    }
  };

  return (
    <Wrapper>
      <Div edge="left">
        <Button onClick={() => handleClick(-1)} disabled={!canPlay}>
          -1
        </Button>
      </Div>
      <Div>
        <Button onClick={() => handleClick(0)} disabled={!canPlay}>
          0
        </Button>
      </Div>
      <Div edge="left">
        <Button onClick={() => handleClick(1)} disabled={!canPlay}>
          +1
        </Button>
      </Div>
      <Popup show={showPopup}>Wrong move</Popup>
    </Wrapper>
  );
};

export default React.memo(Controllers);
