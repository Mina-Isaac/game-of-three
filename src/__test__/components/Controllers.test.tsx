import React from "react";
import ReactDOM from "react-dom";
import configureStore from "redux-mock-store";
import { Provider } from "react-redux";
import { initialState as initialGameState } from "../../store/gameSlice";
import { initialState as initialPlayerState } from "../../store/playerDetailsSlice";
import renderer from "react-test-renderer";
import Controllers from "../../Components/GamePlay/Controllers";
const mockStore = configureStore();
let store: any;

describe("Controllers component should render without issues", () => {
  beforeEach(() => {
    const initialState = { player: initialPlayerState, game: initialGameState };
    store = mockStore(initialState);
  });
  it("renders without crashing", () => {
    const div = document.createElement("div");
    ReactDOM.render(
      <Provider store={store}>
        <Controllers playingAs = {undefined}  />
      </Provider>,
      div
    );
    ReactDOM.unmountComponentAtNode(div);
  });
  it("renders correctly", () => {
    const tree = renderer
      .create(
        <Provider store={store}>
          <Controllers playingAs = {undefined}  />
        </Provider>
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });
});
