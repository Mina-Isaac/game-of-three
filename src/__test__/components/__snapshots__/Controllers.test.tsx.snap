// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Controllers component should render without issues renders correctly 1`] = `
<div
  className="sc-bdfBwQ fMjiPE"
>
  <div
    className="sc-gsTCUz dKvvNN"
  >
    <button
      className="sc-dlfnbm gGNFfv"
      disabled={true}
      onClick={[Function]}
    >
      -1
    </button>
  </div>
  <div
    className="sc-gsTCUz gkedno"
  >
    <button
      className="sc-dlfnbm gGNFfv"
      disabled={true}
      onClick={[Function]}
    >
      0
    </button>
  </div>
  <div
    className="sc-gsTCUz dKvvNN"
  >
    <button
      className="sc-dlfnbm gGNFfv"
      disabled={true}
      onClick={[Function]}
    >
      +1
    </button>
  </div>
  <div
    className="sc-hKgILt jyVaUk"
  >
    Wrong move
  </div>
</div>
`;
